### PR TITLE
Deep Account Analytics in Three Tiers for the Polkadot Data Alliance

### PR DESCRIPTION
**Pull request by sourabhniyogi**
This proposal addresses outstanding [deep account analytics problems for the Web3 Foundation](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/data_analysis_tools.md).
We propose a 3-tier solution:

 - Tier 1. _raw indexing_: Google BigQuery Public Datasets
 - Tier 2. _deep indexing_: shared datasets delivered with Google Analytics Hub
 - Tier 3. _dashboards_: the above indexed data in Apache Superset

This proposal seeks to develop the above solution with the support of a Web3 Foundation grant, but critically addresses 
future maintainance a [Polkadot Data Alliance Bounty](https://docs.google.com/document/d/1ryC6dxcd9tiQsB7KiCc2BY_TwBJ5jKloGfCyVAGCkKo/edit#heading=h.gjdgxs)
that can support work on all 3 tiers with _child bounties_ for Polkadot Data Alliance contributors.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] Payment details have been provided (bank details via email _or_ BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (USDT) address in the application).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

### CODE REVIEW

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the application. I have a few initial questions:
- You seem to have increased the price compared to your previous applications. What's the reason for this?
- If I remember correctly, so far, all your grants have been canceled. Do you think this will be different this time, and why? 
- What are your future plans with this project? Do you plan to apply for treasury funding, etc?

**Comment by sourabhniyogi** [comment]
> Thanks for the application. I have a few initial questions:
> 
> * You seem to have increased the price compared to your previous applications. What's the reason for this?
Whatever we proposed to do here is entirely focused on the specific deep questions being asked by very specific people, but trying to answer them with a different style of work then "oh, its all in state_traceBlock".  So this time we're going to actually get 100% and get 100% deep with the best answer we are capable of doing.  Not a prototype or even 95% of the answer.   We want YOU to use it to answer YOUR questions and I'm happy that the questions are deep and actually important to people.   The previous applications actually required less work, or involved a lot of work but we already did a lot of it.  In this case, there is a lot of serious experimental data work and I did enough causal learning and feature engineering in toy problems / massive advertising datasets to know I'm (a) likely to succeed some with some work (b) unlikely to succeed without some iterations in each pallet (c) it will always need more work to cover the general causal learning problem.  The general causal learning is like (c) but for the specific deep questions its going to be (a)=>(b).    
 
> * If I remember correctly, so far, all your grants have been canceled. Do you think this will be different this time, and why?
Thanks for asking a serious question.  Let me give a serious answer.  This is a bigger problem than causal models.

The first application was about adding polkaholic.io "Verify" at key spots where.  But we were actually looking for a serious maintenance plan for us to build and improve a multichain block explorer.   Unfortunately W3F is not set up for maintenance ("see faq"), so we added some "Verification" features to fit, but no one actually clicked on "Verify", nor  complained about it not working or asked for more -- which is my #1 technique to know that something is valuable.   In contrast, lots of people complained about our XCM indexing, which we consistently improved last Summer/Fall.   

The second application was about zombienet explorer.  I think its a good idea still.  Last December we went deep down into integrating our APIs in polkadot.js (which we love and were disappointed to learn Jacogr wasn't at parity and is remarkably invisible) but in Jan/Feb we found the lack of leadership behind something as simple as delivering polkadot.js apis at console.polkadot.network (for "devs") or an improved version at explorer.polkadot.network  (for the unicorn "Grandma" that uses polkadot lol) linking to various data at analytics.polkadot.network ... there is product management of fundamental UX as a multichain explorer, dev console, powered by multiple data indexers, multiple wallets across desktop/mobile?    This is a serious problem needing serious answers that go beyond Level 1/2 into 5-10 person teams that cost $1.xMM/yr.

After meeting with Parity Data at sub0 and working with Karim+Pavla this Spring that they ALSO have a BigQuery it became clear it would be natural to collaborate on indexing efforts as part of a data collective.  Starting with us+substrate-etl, publishing in Google Public Datasets and continuing into Superset+Looker+Analytics Hub+... tools.  This is far more effective way and doing large amount of useful analysis that isn't like, whatever just happened in the last minute in your account. 

For this third application, it CAN be different if we address the elephant in the room of MAINTENANCE inside that FAQ.   The facts are:
* every successful data project in the blockchain world is going to involve a modern data warehouse like Google Cloud, AWS, or Azure.  We happen to use Google Cloud in a large way and so does Parity
* the W3F does not have a plan for MAINTENANCE.  See FAQ.  
* after a research and prototyping phase, data indexing (/ AI projects ) is all about MAINTENANCE of software, storage, compute, embedded in a large data warehouse
For both this project and the general plan for polkaholic and substrate-etl, we drafted a [Substrate+XCM Parent Bounty](https://docs.google.com/document/d/1ryC6dxcd9tiQsB7KiCc2BY_TwBJ5jKloGfCyVAGCkKo/edit#) support a wide variety of indexing, curated by Parity Data and others with a vested interest in the data.   

Given a 500K DOT/yr budget, we believe this can support 2-3 teams like us in a large way ($1MM/yr) and 5-10 teams in a small way ($100K/yr) that form the core of a "Polkadot Data Alliance".  It is necessary for curators to treat maintenance seriously, have a healthy amount of coopetition, support newcomers, exit incumbents (including us), and require committed to cooperation while also supporting the ability for everyone to build profitable Web3 businesses with  competitive data services.    Many people have a misguided idea that this bounty should about providing a means to support a single multichain block explorer like Subscan, and its disturbing when I see people who purport to be proponents of decentralized offer that as a goal.   We instead think the future is much brighter than that and that people will do a lot more at the edges, and then act on that data.  The very nature of app chains / pallets / ..  demands a far decentralized form of engineering than imagining that one team can properly support everything.

We believe W3F's model forms a good template for the Polkadot Data Alliance's parent bounty to follow, but the software enhancements and indexing strategy must be explicitly done differently.   

I have requested a discussion between the RFP requesters here (Michalis + Diogo) and Parity Data (Karim+Pavla) to see how we can move forward and chart a course together, not just for this proposal, but everyone else facing constraints and wanting to have an impact for our every data processing lives.  

> * What are your future plans with this project? Do you plan to apply for treasury funding, etc?
The [Substrate+XCM Parent Bounty](https://docs.google.com/document/d/1ryC6dxcd9tiQsB7KiCc2BY_TwBJ5jKloGfCyVAGCkKo/edit#) requires 5 parent bounty curators and each child bounty requires some curator to be identified.   We request that W3F to be a parent curator and someone who uses project (e.g Michalis) to be a child bounty curator, after this project starts working well enough for them.  

With a serious plan for maintenance, our hope is that we can develop a analytics.polkadot.network ( or https://analytics.web3.foundation/ ) shared place where everyone can contribute multiple shared public datasets and build+share dashboards.   It helps to have a set of curators, a few core teams contributing a lot and dozens of teams contributing very special deep ways based on their deep knowledge.    We have 2 dozen ideas of child bounties, but we do not have the patience to turn all of them into W3F proposals but our hope is others can and will do whatever they do best within this sort of structure.   I do not want to have a company where every engineer needs a proposal manager to work with a dozen foundations to cover every line item.  

We hope we have a like a few deep questions from you where we can deliver a few of them thereafter within this general place, and have it be a place where we can grow a business and help others grow theirs doing whatever we are all good at.

**Review by michalisFr — APPROVED** [review]
I think after the discussion we had with the Parity Data team and the changes @sourabhniyogi made to the proposal, that this is a reasonable approach and proposal. It answers our specific questions while setting the groundwork for how future data-related requests should be treated.

In short, what we discussed yesterday involved a hybrid funding model where development work (including UI) for specific needs of W3F would be covered by grants. This might include data engineering work needed to get the necessary data. This type of work is expected to be reduced for future requests/projects, since it builds on the work already done and as more teams get involved. 

On the other hand, work on raw data, data engineering work that is broadly beneficial/needed, and maintenance costs (mostly storage costs) would be covered by the Polkadot Data Alliance bounty.

So, overall I think this is a good proposal, but since I'm not part of the grants team I'm looking forward to see @dsm-w3f's feedback, who was also on the call.

**Comment by dsm-w3f** [comment]
@sourabhniyogi thanks for updating the proposal. Seems that @michalisFr gives his ok from the scope perspective for this demand. I just notice that the deliverable 0c. the testing guide is not present. Could you please add it? If the automated tests don't make sense for your application at least a manual testing guide with the steps for using the application, with examples including expected inputs and outputs should be provided. Without this, we are not able to evaluate the deliverables.

Furthermore, the price still looks high to me. There are many man-hours included in six months roadmap, which looks large for the scope consdered. Considering @Noc2 concerns about not actually having delivered previous grants, I think would be easier to approve this proposal if you go for a level 2 grant.

**Review by dsm-w3f — CHANGES_REQUESTED** [review]
See comment https://github.com/w3f/Grants-Program/pull/1768#issuecomment-1587864575

**Comment by sourabhniyogi** [comment]
> @sourabhniyogi thanks for updating the proposal. Seems that @michalisFr gives his ok from the scope perspective for this demand. I just notice that the deliverable 0c. the testing guide is not present. Could you please add it? If the automated tests don't make sense for your application at least a manual testing guide with the steps for using the application, with examples including expected inputs and outputs should be provided. Without this, we are not able to evaluate the deliverables.
> 
> Furthermore, the price still looks high to me. There are many man-hours included in six months roadmap, which looks large for the scope consdered. Considering @Noc2 concerns about not actually having delivered previous grants, I think would be easier to approve this proposal if you go for a level 2 grant.

I reduced Milestone 2  and adjusted to Level 2 as per your request by lowering trace storage systematization work and proxy/era treatment [in expectation this will be covered by PDA child bounties].

File: applications/DeepAccountAnalytics-PolkadotDataAlliance.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+This proposal is in response to the RFP [Data Analysis Tools for Substrate-based Blockchains](https://github.com/w3f/Grants-Program/blob/master/docs/RFPs/Open/data_analysis_tools.md)
+which seeks to answer deep account analytics questions of:
+
+**Account References**:
+* Which transactions/accounts were responsible for the reserved balance in an account?
+* What modules currently depend on consumers, providers, and sufficients reference counters for a certain account, and which transactions introduced/removed those references?
+* What are the pallets responsible for reserves/holds and locks/freezes on an account?
+
+**Democracy**:
+* Which accounts have delegated OpenGov votes to an account or to which accounts the account in question has delegated their votes to for each track,  taking into account indirect delegations too (e.g. Account A delegates to Account B which delegates to Account C)?
```
[the remark this answers is not in the document]
Comment (sourabhniyogi) [reply applications/DeepAccountAnalytics-PolkadotDataAlliance.md]: covered!

**Review by semuelle — APPROVED** [review]
> what we discussed yesterday involved a hybrid funding model where development work (including UI) for specific needs of W3F would be covered by grants.

> It answers our specific questions while setting the groundwork for how future data-related requests should be treated.

Sounds good to me. Happy to proceed.

**Review by dsm-w3f — APPROVED** [review]
@sourabhniyogi thank you for adding my requests to the application. I notice that the size of the index was reduced for this grant, but this is not a problem since the main purpose is to develop the software. I'm happy to go forward with it and support this project.

**Comment by sourabhniyogi** [comment]
Thank you all very much

**Comment by keeganquigley** [comment]
Hi @sourabhniyogi how is milestone 1 coming along? Can you provide any updates?

**Comment by sourabhniyogi** [comment]
We did our backend refactorization this summer extending to traces, did a
PoC internally and are awaiting for a Polkadot Data Alliance curator set to
materialize for this project to have long term meaning.   The last word
from the others on the PDA is "the process is painfully slow and we're
still in internal alignments 🙁 We're doing our best."

**Pull request merged**